src/main/kotlin/org/rust/cargo/project/workspace/impl/CargoProjectWorkspaceImpl.kt

```diff
@@ -107,30 +100,42 @@ class CargoProjectWorkspaceImpl(private val module: Module) : CargoProjectWorksp
      *            until the `Cargo.toml` becomes sound
      */
     override val projectDescription: CargoProjectDescription?
-        get() = lock (lock) { cached }
+        get() {
+            check(ApplicationManager.getApplication().isReadAccessAllowed)
+            return cached
+        }
 
-    private fun notifyCargoProjectUpdate(r: UpdateResult) {
+    private fun commitUpdate(r: UpdateResult) {
         ApplicationManager.getApplication().assertIsDispatchThread()
+        if (module.isDisposed) return
+
+        runWriteAction {
+            if (r is UpdateResult.Ok) {
+                cached = r.projectDescription
+                updateModuleDependencies(r.projectDescription);
+            }
+            notifyCargoProjectUpdate(r)
```

Comment:
BTW, a short post-mortem: initial decision of taking _global WL_ holding local lock was smelly as hell, and should not be introduced in the future -- it's a source of the _guaranteed_ dead-lock, since (most of) the actions are performing under RL.